eggplant/market/models/cart.py

```diff
@@ -53,35 +60,49 @@ def add_to_items(self, product=None, quantity=1, delivery_date=None):
 
     def remove_from_items(self, product=None, quantity=1, delivery_date=None):
         current = self.items.filter(product=product,
-                                    delivery_date=delivery_date)
-        if current.exists():
-            if current[0].quantity - quantity > 1:
-                current.update(quantity=models.F('quantity') - quantity)
+                                    delivery_date=delivery_date).first()
+        if current:
+            if current.quantity - quantity > 1:
+                current.quantity -= quantity
+                current.save()
             else:
                 self.items.filter(basket=self, product=product, delivery_date=delivery_date).delete()
 
     def get_total_amount(self):
-        # TODO: This should return a Money type instead of decimal
-        # TODO: When calculating this, we should fail when currencies differ
-        total = Decimal('0')
-        for item in self.items.all():
-            total += item.quantity * item.product.price.amount
-        return total
+
+        # FIXME this check should be removed, it is better to check for the currenices when an item is added
+        # sanity check for the currencies
+
+        if 1 != self.items.all().values('product__price_currency').annotate(total=Count('product__price_currency')).count():
+            raise ValidationError('The products have different currencies')
+
+        total = self.items.all().aggregate(total=Sum(F('product__price') * F('quantity')), cur=Max('product__price_currency'))
+        return Money(total['total'], total['cur'])
 
     def get_items_count(self):
         return self.items.all().count()
 
     @transaction.atomic
     def do_checkout(self):
-        # TODO: Why on earth are we importing this here
-        from .payment import Payment
-        Payment.objects.create(
+        payment = Payment.objects.create(
             amount=self.get_total_amount(),
+            # FIXME: get valid account
```

Comment:
It would add more consistency to the shopping logic. Everything is based on the `account`. It it actually quite nice task and it shouldn’t be very difficult. Only someone has to decide where the selector for the active account will be. If someone less experience decide to do, I'm happy to assist him/her. 

When we talk about shopping logic the payment model it is not optimal as well. It should be called an `order`. Order can have many `payments` and many `orderItems`. As it is now it has many limitation for real-life situations.